src/System.Private.CoreLib/shared/System/Number.BigInteger.cs

```diff
@@ -0,0 +1,749 @@
+                }
+
+                return quotient;
+            }
+
+            public static uint LogBase2(uint value)
+            {
+                Debug.Assert(value != 0);
+                return BitScanReverse(value);
+            }
+
+            public static uint LogBase2(ulong value)
+            {
+                Debug.Assert(value != 0);
+
+                uint upper = (uint)(value >> 32);
+
+                if (upper != 0)
+                {
+                    return 32 + LogBase2(upper);
+                }
+
+                return LogBase2((uint)(value));
+            }
+
+            public static void Multiply(ref BigInteger lhs, uint value, ref BigInteger result)
+            {
+                if (lhs.IsZero() || (value == 1))
+                {
+                    result.SetValue(ref lhs);
+                    return;
+                }
+
+                if (value == 0)
+                {
+                    result.SetZero();
+                    return;
+                }
+
+                int lhsLength = lhs._length;
+                int index = 0;
+
+                ulong carry = 0;
+
+                while (index < lhsLength)
+                {
+                    ulong product = ((ulong)(lhs._blocks[index]) * value) + carry;
+                    carry = product >> 32;
+                    result._blocks[index] = (uint)(product);
+
+                    index++;
+                }
+
+                if (carry != 0)
+                {
+                    Debug.Assert(unchecked((uint)(lhsLength)) + 1 <= MaxBlockCount);
+                    result._blocks[index] = (uint)(carry);
+                    result._length += (lhsLength + 1);
+                }
+            }
+
+            public static void Multiply(ref BigInteger lhs, ref BigInteger rhs, ref BigInteger result)
+            {
+                if (lhs.IsZero() || rhs.IsOne())
+                {
+                    result.SetValue(ref lhs);
+                    return;
+                }
+
+                if (rhs.IsZero())
+                {
+                    result.SetZero();
+                    return;
+                }
+
+                ref readonly BigInteger large = ref lhs;
+                int largeLength = lhs._length;
+
+                ref readonly BigInteger small = ref rhs;
+                int smallLength = rhs._length;
+
+                if (largeLength < smallLength)
+                {
+                    large = ref rhs;
+                    largeLength = rhs._length;
+
+                    small = ref lhs;
+                    smallLength = lhs._length;
+                }
+
+                int maxResultLength = smallLength + largeLength;
+                Debug.Assert(unchecked((uint)(maxResultLength)) <= MaxBlockCount);
+
+                // Zero out result internal blocks.
+                Buffer.ZeroMemory((byte*)(result._blocks.GetPointer()), (MaxBlockCount * sizeof(uint)));
```

Comment:
Did you mean to zero `maxResultLength` only?

Reply:
This was copied from the native code: https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/bignum.cpp#L451

Just zeroing out `maxResultLength` is probably better, however.

Reply:
Fixed.